kinesis/src/main/scala/akka/stream/alpakka/kinesis/KinesisSourceStage.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.kinesis
+
+import java.util.concurrent.Future
+
+import akka.actor.ActorRef
+import akka.stream.alpakka.kinesis.KinesisSourceStage._
+import akka.stream.alpakka.kinesis.{KinesisSourceErrors => Errors}
+import akka.stream.stage.GraphStageLogic.StageActor
+import akka.stream.stage._
+import akka.stream.{Attributes, Outlet, SourceShape}
+import com.amazonaws.AmazonWebServiceRequest
+import com.amazonaws.handlers.AsyncHandler
+import com.amazonaws.services.kinesis.AmazonKinesisAsync
+import com.amazonaws.services.kinesis.model._
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object KinesisSourceStage {
+
+  private[kinesis] final case class GetShardIteratorSuccess(result: GetShardIteratorResult)
+
+  private[kinesis] final case class GetShardIteratorFailure(ex: Throwable)
+
+  private[kinesis] final case class GetRecordsSuccess(records: GetRecordsResult)
+
+  private[kinesis] final case class GetRecordsFailure(ex: Throwable)
+
+  private[kinesis] final case object Pump
+
+}
+
+class KinesisSourceStage(shardSettings: ShardSettings, amazonKinesisAsync: => AmazonKinesisAsync)
+    extends GraphStage[SourceShape[Record]] {
+  private val out = Outlet[Record]("Records")
+
+  override def shape: SourceShape[Record] = new SourceShape[Record](out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with StageLogging {
+
+      import shardSettings._
+
+      private[this] var iterator: String = _
+      private[this] var buffer = mutable.Queue.empty[Record]
+      private[this] var self: StageActor = _
+
+      override def preStart(): Unit = {
+        self = getStageActor(awaitingShardIterator)
+        getShardIterator()
+      }
+
+      setHandler(shape.out, new OutHandler {
+        override def onPull(): Unit = self.ref ! Pump
+      })
+
+      private def awaitingShardIterator(in: (ActorRef, Any)): Unit = in match {
+        case ((_, GetShardIteratorSuccess(result))) => {
+          iterator = result.getShardIterator
+          self.become(awaitingRecords)
+          requestRecords()
+        }
+        case ((_, GetShardIteratorFailure(_))) => {
+          log.error("Failed to get a shard iterator for shard {}", shardId)
+          failStage(Errors.GetShardIteratorError)
+        }
+        case ((_, Pump)) => ()
+      }
+
+      private def awaitingRecords(in: (ActorRef, Any)): Unit = in match {
+        case ((_, GetRecordsSuccess(result))) => {
+          val records = result.getRecords.asScala
+          Option(result.getNextShardIterator).fold {
+            log.info("Shard {} returned a null iterator and will now complete.", shardId)
+            completeStage()
+          } { iterator = _ }
+          if (records.size > 0) {
+            records.foreach(buffer.enqueue(_))
```

Comment:
I am a bit wary about this unbounded buffer. I see that `GetRecordsRequest` supports specifying a limit of how many records to get. With that we could batch up the results. One complication with this approach is that a sequence shard iterator type then needs to be used.

Do you think batching results could be doable?